### PR TITLE
Refactor the JSON-RPC connection

### PR DESCRIPTION
**Pull request by sylane**
### Changed

- The name of the grisp_connect configuration key to control the timout of
individual JSON-RPC requests changed from ws_requests_timeout ot
ws_request_timeout.
- Le default log filter changed to trying to filter out only some messages to
filtering out all progress messages, as it wasn't working reliably.
- The connection is not a persistent process anymore, it is now a transiant
process handling a connection and dying when the connection is closed.
- Internally, the JSON-RPC is parsed into a list of atom or binaries to pave the
road for namespaces. foo.bar.Buz is parsed into [foo, bar, <<"Buz">>] (if foo
and bar are already existing atoms, but 'Buz' is not).

## Fixed

- The client is now waiting 1 second before trying to reconnect when it gets
disconnected fomr the server.

### CODE REVIEW

**Comment by maehjam** [comment]
Can you explain to me the different purposes of the modules `grisp_connect_connection`, `grisp_connect_client` and `grisp_connect_jsonrpc`. First impression is that it's all mixed up. This needs better documentation.

**Comment by sylane** [comment]
@maehjam It is less mixed up than it was, but I get that it may still be under documented. grisp_connect_jsonrpc is only doing JSON-RPC encoding/decoding, grisp_connect_connection encapsulate a JSON-RPC connection, and grisp_connect_client encapsulate the girps_connect client protocol.

File: src/grisp_connect_connection.erl
Comment (maehjam) [inline src/grisp_connect_connection.erl:1]: I feel with the timeouts and the distinction between connected and disconnected in `handle_call` this would be much better readable as a `gen_statem`. The module name is a bit strange, I liked `grisp_connect_ws` better.

**Review by maehjam — COMMENTED** [review]
1. The logger changes and architecture changes of the websocket connection shouldn't be in the same PR. Both are changes in parts that are hard to test. They each need very careful testing and probably manual testing in addition to the automatic testing.
2. Is this working with the current `grisp_manager` implementation or do we need to adjust `grisp_manager` as well?

**Comment by peerst** [comment]
This PR is stuck for a while now.  Adding 1860 Lines and removing 430 is clearly beyond the size of reviewability (max 500 lines).  How do we want to proceed with this?  @sylane can you split it up in several PRs?

Also from what I can see so far: it looks in parts more like a rewrite than a refactoring.  I'm not sure if that is due to too many refactoring steps stacked on top of each other so it only looks like a rewrite?  

Rewriting or large refactors have a tendency to sneak in new bugs, so therefore require substantial improvement of testing at the same time: is that the case here?

How is the relationship between this code and DAB?  Will this eventually end up in DAB?  Parts of it?  I don't want to have to maintain two forks of the same thing if that is the actual case.

**Comment by sylane** [comment]
@peerst This is already only some part of a bigger changeset that I split so it was simpler to review... Some changes already got merged in, but there is still more changes pending. The changes here are to go toward having a self-contained independent json-rpc client instead of having json-rpc logic spread everywhere in grisp-connect. So the connection is indeed a rewrite, but all the rest are changes to mostly remove things that are now abstracted by the new connection. I don't think I could easily split the changes further. For the testing, I wrote more tests, and I made the code more test-friendly, as the json-rpc connection can be tested completely independently. If we want more tests for it, I could write more, this shouldn't be complicated.

File: src/grisp_connect_connection.erl [hunk 1]
```diff
@@ -0,0 +1,741 @@
...
+  | {conn, pid(), {remote_error, Code :: integer() | atom(), Message :: undefined | binary(), Data :: term(), Ctx :: term()}}
+  | {conn, pid(), {local_error, Reason:: atom(), Ctx :: term()}}.
+
+-export_type([error_mapping/0, method/0, handler_messages/0]).
+
+
+%--- Macros --------------------------------------------------------------------
+
+-define(ENABLE_TRACE, false).
+-if(?ENABLE_TRACE =:= true).
```
Comment (ziopio) [inline src/grisp_connect_connection.erl]: So is it possible to step over written definitions?

File: src/grisp_connect_client.erl [hunk 2]
```diff
@@ -183,22 +237,112 @@ handle_common(info, Info, State, Data) ->
...
+    Data#data{conn = undefined}.
+
+-spec conn_post(data(), jarl:method(), atom(), map(),
+                undefined | on_result_fun(), undefined | on_error_fun())
+    -> data().
+conn_post(Data = #data{conn = Conn}, Method, Type, Params, OnResult, OnError)
+  when Conn =/= undefined ->
+    ReqCtx = #{on_result => OnResult, on_error => OnError},
+    Params2 = maps:put(type, Type, Params),
+    case jarl:post(Conn, Method, Params2, ReqCtx) of
```
Comment (ziopio) [inline src/grisp_connect_client.erl:332]: Why do you have` jarl:request` and` jarl:post` and you only use `jarl:post` Json RPC does not have the concept of a `post` but only request/notify

Comment (ziopio) [reply src/grisp_connect_client.erl:332]: Also, in the jarl application you have this 700+ lines `gen_statem` that implements JsonRPC message handling and GUN.
Could these 2 concepts be split in 2 modules in Jarl,as it was in `grisp_connect` before this PR?
The jarl_connection module looks already too big

**Review by ziopio — CHANGES_REQUESTED** [review]

File: src/grisp_connect.app.src [hunk 3]
```diff
@@ -21,26 +21,24 @@
...
-           }}
+            % Enable our own default handler,
+            % which will receive all events from boot
+            {handler,
+                grisp_connect_log_handler,
+                grisp_connect_logger_bin,
+                #{formatter => {grisp_connect_logger_bin, #{}},
+                filter_default => log,
+                filters => [
+                    {disable_progress, {fun logger_filters:progress/2, stop}}
```
[the remark this answers is not in the document]
Comment (sylane) [reply src/grisp_connect.app.src:39]: I am not 100% sure that it make things better. It looked better for me, and this is the common solution I found about our problem with logging certificate, but others should test it too.

File: src/grisp_connect_connection.erl (hunk 1 above)
Comment (sylane) [reply src/grisp_connect_connection.erl]: @ziopio Not sure what you mean by that...

File: src/grisp_connect_client.erl (hunk 2 above)
Comment (sylane) [reply src/grisp_connect_client.erl:332]: As the idea was first to consolidate the JSON-RPC in a single place in grisp_connect first, this is what I did with jarl_connection, but then I moved it out of grisp_connect to make the PR more compact. Maybe this could be refactored again to split again, not sure if it is needed now though, I am open to it. The main reason I think it was more important before is because the guns management and the json-rpc was done in different processes before, now it is done in the same process.

Comment (sylane) [reply src/grisp_connect_client.erl:332]: @maehjam What do you think about that @maehjam ?

File: src/grisp_connect_connection.erl (hunk 1 above)
Comment (ziopio) [reply src/grisp_connect_connection.erl]: You define `ENABLE_TRACE = false` and then use it right away to check if it is true. Without context this looks like the if section will never apply.

File: src/grisp_connect_client.erl (hunk 2 above)
Comment (ziopio) [reply src/grisp_connect_client.erl:332]: @sylane still... why `post` instead of `request` ?

Comment (maehjam) [reply src/grisp_connect_client.erl:332]: I'm also confused what `post` is. There is no documentation in `jarl` about it.

Comment (sylane) [reply src/grisp_connect_client.erl:332]: post it to asynchronously post a request, request is doing a post and wait for the response so it is synchronous

Comment (sylane) [reply src/grisp_connect_client.erl:332]: When using post, you need to handle the asynchronous messages with either a response or an error.

**Comment by maehjam** [comment]
Sorry, I still don't think I'm able to review this. It is too much changes at once. Now, we have changes here and a new app `jarl` to review together. Splitting it up, didn't reduce the difficulty to review.

I feel before we can review the changes here, we need to review `jarl` which is 2000+ lines in one commit without any documentation. I looked into it and it's really hard to read. For example a function named `jarl:error/5` can do different things: Did we receive a json-rpc error from the server or do we want to send one or something else? And what is the difference between `jarl:error/5` and `jarl:reply/2`, when `error` is about replying an error? The naming of `jarl:post/4` was already discussed inline. The name `post` is usually associated with HTTP and has no equivalent in JSON-RPC.

Furthermore, I have to note: Even when we move a big chunk of this rewrite in a separate library that is still a big rewrite. We already have two JSON-RPC client implementations here in `grisp_connect` as well as in `dab` and should reuse the knowledge and preferably also the implementations. I like the idea of a reusable JSON-RPC client app, but that should be developed together with the people who worked on the implementations in `grisp_connect` and `dab` to use their experience. Just putting us here as reviewers after implementing 2000+ lines doesn't work. That is too much to read while trying to compare with the existing work (without readable diffs).

If we want this big rewrite. I would prefer that we first develop `jarl` starting with the concept and general architecture that we discuss and agree on as a team. Then we can review the implementation in smaller steps (commits). 

We should make sure a library like `jarl` is usable by `grisp_connect` as well as by `dab`, because otherwise what is the point of a separate app? Can you tell me whether `jarl` can be used in `dab` as well?

**Comment by sylane** [comment]
This PR is only around 330+330- lines if you don't count tests, docs and config, it is big but I don't think it is unrasonable. I understand that moves some review work to jarl, but it is now a self-contains library with limited scope and API. They weren't any documentation with the old jrpc-code, at least now theres is well defined type specifications. I could add more documentation, this is not a problem, but would that change your mind ?
API jarl:error/5 is not doing different things, it is sending an error to the connection, either in the context of a request or not, I am not sure what you mean by it is doing different things. Everything you receive from the json-rpc peer, you get either synchronously when using jarl:request, or asynchronously with messages definied in jarl type specs. Post is used in http but convei some meaning too, it as a strong meaning of asynchronicity this is why it is used there, but this is just naming, if this was the only problem it would be easy to resolve, right ?
I am curious about what would be the smaller steps to implement a json-rpc client library, and I don't see why this couldn't be used by DAB, we could discuss that if you were interested.
Anyway, I am not sure how to react to your comment, the feeling is that there is nothing to do in your opinion to get this to a mergable state, you have issue with some naming and with documentation, all good points, but I don't feel that would change your opinion on the PR... Should I just delete the PR and the branch and forget about it ?

**Comment by ziopio** [comment]
@sylane  I think we should discuss jarl usage, architecture and API. This PR would be ok but it relie on us accepting jarl as is.
The best would be:
- you present jarl to the team
- we try to understand if and how the design would also help DAB or be improved using DAB design
- We agree on an API that can be generic and reusable ecc
- We define what needs to change from what jarl is now.
- You rewrite Jarl with incremental PRs so we can quickly follow the development.
So the PR can be parked here until we are fine with jarl.

**Comment by maehjam** [comment]
Just to clear it up: The examples I mentioned about the function naming `error` and `post` in `jarl` were just to point out one of the things that make it difficult to read the code. The names were not self-explanatory to me, and like that I find most of the code hard to understand. The bigger issue is that it's just too much at once. Since it's one big commit, I can't even follow your steps of implementation to get the idea und structure that is probably totally clear to you as the author.

Thanks @ziopio for the proposal on how we can continue here. That is for sure one way. I would like a decision from @peerst and the rest of the team on how we proceed, before we put too much time into it. That can be discussed after the first point of presenting `jarl` to the team.

File: docs/grisp_connect_architecture.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+It is not supervised, the process starting it must monitor it.
+
+It provides a high-level API to a JSON-RPC connection:
+
+ - Perform synchronous requests
+ - Start asynchronous requests
+ - Reply to a request
+ - Send and error result for a request
+ - Send asynchronous notifications
+ - Send generic errors
```
Comment (maehjam) [inline docs/grisp_connect_architecture.md:64]: That formulation is a bit unclear. Do you mean the automatic sending for predefined JSON-RPC errors from the JSON-RPC specifications? Then instead of "generic errors" rephrase it to "predefined JSON-PRC errors".

File: docs/grisp_connect_architecture.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+When performing an asynchronous request, the caller can give an opaque context
+term, that will given back when receiving a response or an error for this
```
Comment (maehjam) [inline docs/grisp_connect_architecture.md:67]: "that will <ins>be</ins> given back"

File: config/dev.config [hunk 6]
```diff
@@ -10,18 +10,16 @@
...
             {handler, default, logger_std_h, #{
                 config => #{type => standard_io},
+                filter_default => log,
                 filters => [
-                    % Filter out supervisor progress reports so
-                    % TLS certificates are not swamping the console...
-                    {filter_out_progress, {
-                        fun grisp_connect_logger_bin:filter_out/2,
-                        {supervisor, report_progress}}}
+                    {disable_progress, {fun logger_filters:progress/2, stop}}
```
Comment (maehjam) [inline config/dev.config:19]: It would be good to keep the comment why we do that here and also add it to `local.config`, `test.config`.

File: src/grisp_connect_api.erl [hunk 7]
```diff
@@ -1,152 +1,76 @@
...
+%--- API -----------------------------------------------------------------------
 
-% #doc Assembles a jsonrpc notification
--spec notify(Method :: atom() | binary(),
-              Type :: atom() | binary(),
-              Params :: map()) -> Encoded :: binary().
-notify(Method, Type, Params) ->
-    Rpc = {notification, Method, maps:put(type, Type, Params)},
-    grisp_connect_jsonrpc:encode(Rpc).
+% @doc Handles requests, notifications and errors from grisp.io.
```
Comment (maehjam) [inline src/grisp_connect_api.erl:15]: I see it only handling notifications and requests, not errors.

File: src/grisp_connect_api.erl [hunk 8]
```diff
@@ -1,152 +1,76 @@
...
     try
         URL = maps:get(url, Params),
-        Reply = case grisp_connect_updater:start_update(URL) of
+        case grisp_connect_updater:start_update(URL) of
             {error, grisp_updater_unavailable} ->
-                {error, -10, grisp_updater_unavailable, undefined, ID};
+                {error, grisp_updater_unavailable, undefined, undefined, ID};
             {error, already_updating} ->
-                {error, -11, already_updating, undefined, ID};
+                {error, already_updating, undefined, undefined, ID};
```
Comment (maehjam) [inline src/grisp_connect_api.erl:40]: This error is not defined in `grisp_connect_client`.

File: src/grisp_connect_api.erl [hunk 9]
```diff
@@ -1,152 +1,76 @@
...
-handle_request(?method_post, #{type := <<"start_update">>} = Params, ID) ->
+    {reply, Info, ID};
+handle_request([?method_post], #{type := <<"start_update">>} = Params, ID) ->
     try
         URL = maps:get(url, Params),
-        Reply = case grisp_connect_updater:start_update(URL) of
+        case grisp_connect_updater:start_update(URL) of
             {error, grisp_updater_unavailable} ->
-                {error, -10, grisp_updater_unavailable, undefined, ID};
+                {error, grisp_updater_unavailable, undefined, undefined, ID};
```
Comment (maehjam) [inline src/grisp_connect_api.erl:38]: All the error tuples are not handled as return values is `grisp_connect_client:handle_connection_message/2`.

**Review by maehjam — CHANGES_REQUESTED** [review]
I reviewed the first 8 files and found that the JSON-RPC API must be broken, since errors are not handled. I was wondering whether the tests wouldn't catch that. Please extend `grisp_connect_api_SUITE` to better cover the API. It only tests two methods.

I stop my review here till the tests pass.

There are already some change requests in the inline comments. 

One other thing I noticed is that `rebar.lock` is missing. With the new `jarl` library as a dependency it should be updated.

File: CHANGELOG.md [hunk 10]
```diff
@@ -8,6 +8,23 @@ and this project adheres to
...
+filtering out all progress messages, as it wasn't working reliably.
+- JSON-RPC logic was extracted into the jarl library.
+- Jarl parses the methods into a list of atom or binaries to pave the
+road for namespaces. foo.bar.Buz is parsed into [foo, bar, <<"Buz">>] (if foo
+and bar are already existing atoms, but 'Buz' is not).
+
+## Fixed
+
+- The client is now waiting 1 second before trying to reconnect when it gets
+disconnected from the server.
```
[the remark this answers is not in the document]
Comment (sylane) [reply CHANGELOG.md:26]: It is done in in grisp_connect_client, in the connecting function.

[the remark this answers is not in the document]
Comment (maehjam) [reply CHANGELOG.md:26]: I did look again and found it in `waiting_ip`. So that is a counter intuitive place to do it. I was expecting it in `connecting` somewhere, for example in `enter` there. Is there some deeper reason to put it in `waiting_ip` that I don't see?

File: src/grisp_connect_api.erl (hunk 7 above)
Comment (sylane) [reply src/grisp_connect_api.erl:15]: It wasn't before the refactor IIRC, so I didn't add it.

File: CHANGELOG.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (sylane) [reply CHANGELOG.md:26]: It doesn't make much sense to not check again if the IP is available after a connection error, so going back to waiting_ip seems better, and as we want to wait between each ip checks, better leveraging this than behaving differently if we were connected or not before then.

File: src/grisp_connect_api.erl (hunk 7 above)
Comment (sylane) [reply src/grisp_connect_api.erl:15]: In fact, the request errors are handled in grisp_cllient: https://github.com/grisp/grisp_connect/blob/sylane/cleanup-connection/src/grisp_connect_client.erl#L260

Comment (maehjam) [reply src/grisp_connect_api.erl:15]: It was done before the refactoring (you can clearly see it in the diff) and it had a totally different documentation.
So that documentation is about your change and it doesn't fit. Fix it.

File: CHANGELOG.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (maehjam) [reply CHANGELOG.md:26]: But it's not leveraged. The waiting between `waiting_ip` is implemented in line 156 and the waiting between connects in line 144. They are not related at all.

File: src/grisp_connect_api.erl (hunk 7 above)
Comment (sylane) [reply src/grisp_connect_api.erl:15]: You are right, and I remembered that and posted the second comment about where the error is handled. Now the error are handled in a generic way in grisp_connect_client, line 260

File: CHANGELOG.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (sylane) [reply CHANGELOG.md:26]: It is, because by switching to state `waiting_ip` from either the `connecting` or `connected` state you will have the correct delay. Maybe using `repeat_state` instead of `next_state` at line 156 and not registering a new `state_timeout` would make it clearer for you ? This way the timeout would only be setup in `waiting_ip` enter function.

File: test/grisp_connect_log_SUITE.erl [hunk 11]
```diff
@@ -187,6 +188,23 @@ last_seq() ->
     send_jsonrpc_result(#{seq => Seq, dropped => 0}, Id),
     Seq.
 
+reset_log() ->
+    reset_log(undefined).
+
+reset_log(LasSeq) ->
```
[the remark this answers is not in the document]
Comment (maehjam) [reply test/grisp_connect_log_SUITE.erl:194]: LasSeq -> LastSeq

**Review by maehjam — APPROVED** [review]
🚀

**Review by ziopio — APPROVED** [review]
